### PR TITLE
support multiple flavor of notebooks

### LINKED ISSUE #12157: Add support to provide multiple flavor of deployment notebooks
Requested by @SQLvariant, he would like to contribute PowerShell notebook for the deployment scenarios, make it easier for the PowerShell users to understand and troubleshoot the deployment notebooks.

### PR DESCRIPTION
**Pull request by alanrenmsft**
This PR fixes #12157

This enables @SQLvariant to contribute PowerShell notebooks for our deployment scenarios.
this is only enabled for configuration + notebook based dialog deployment targets, since wizards are more complicated and even generate python code cell and add to notebook at runtime.

This is how it looks at runtime, a notebook type selector is automatically added to the end of the dialog if multiple notebooks are provided for that deployment option.

image

@SQLvariant These are all the deployment options:
image

you can contribute PowerShell notebooks for the follow scenarios:
1. SQL Server container image - 2 notebooks: https://github.com/microsoft/azuredatastudio/blob/main/extensions/resource-deployment/package.json#L96 and https://github.com/microsoft/azuredatastudio/blob/main/extensions/resource-deployment/package.json#L147
1. Azure SQL Edge - 5 notebooks: https://github.com/microsoft/azuredatastudio/blob/main/extensions/asde-deployment/package.json#L62, all the notebooks in the extension can have PS version

### How to provide additional notebooks:
Replace the notebook value in the package.json

```JSON
              "notebook": "./notebooks/edge/deploy-sql-edge-local.ipynb",
```

with

```JSON
              "notebook": [
                {
                  "type": "Python",
                  "path": "./notebooks/edge/deploy-sql-edge-local.ipynb"
                },
                {
                  "type": "Powershell",
                  "path": "./notebooks/edge/deploy-sql-edge-local-PS.ipynb"
                }
              ],
```

### CODE REVIEW

File: extensions/resource-deployment/src/ui/deploymentInputDialog.ts [hunk 1]
```diff
@@ -81,7 +102,10 @@ export class DeploymentInputDialog extends DialogBase {
 			if (this.dialogInfo.runNotebook) {
 				this.executeNotebook(this.dialogInfo);
 			} else {
-				this.notebookService.launchNotebook(this.dialogInfo.notebook).then(() => { }, (error) => {
+				const notebook = Array.isArray(this.dialogInfo.notebook) ?
+					this.dialogInfo.notebook.find(nb => nb.type === model.getStringValue(NotebookTypeVariableName))?.path :
+					this.dialogInfo.notebook;
+				this.notebookService.launchNotebook(notebook!).then(() => { }, (error) => {
```
Comment (aaomidi) [inline extensions/resource-deployment/src/ui/deploymentInputDialog.ts:108]:
```				
this.notebookService.launchNotebook(notebook!).catch((error) => {
					vscode.window.showErrorMessage(error);
				});
```

File: extensions/resource-deployment/src/ui/deploymentInputDialog.ts [hunk 2]
```diff
@@ -91,6 +115,6 @@ export class DeploymentInputDialog extends DialogBase {
 	}
 
 	private executeNotebook(notebookDialogInfo: NotebookBasedDialogInfo): void {
-		this.notebookService.backgroundExecuteNotebook(notebookDialogInfo.taskName, notebookDialogInfo.notebook, 'deploy', this.platformService);
+		this.notebookService.backgroundExecuteNotebook(notebookDialogInfo.taskName, <string | NotebookPathInfo>notebookDialogInfo.notebook, 'deploy', this.platformService);
```
Comment (aaomidi) [inline extensions/resource-deployment/src/ui/deploymentInputDialog.ts:118]: <> is deprecated, use as

```
(notebookDialogInfo.notebook as string | NotebookPathInfo)
```

**Review by Zenix27 — APPROVED** [review]

**Comment by ranasaria** [comment]
I am curious about the business need for multiple notebooks for the same deployment sicenario. Is this because we expect the end users to understand and extend the notebooks as per their own need?

**Comment by SQLvariant** [comment]
@ranasaria that is correct.  The way the Notebooks currently come out, they look like incomprehensible jibberish to many customers.  If they get an error, they might have no idea where to start troubleshooting.

**Comment by ranasaria** [comment]
@SQLVariant, I think type may not be best way to distinguish onenotebook from another. I think a short description string might be more appropriate.

File: extensions/resource-deployment/src/ui/deploymentInputDialog.ts (hunk 2 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/deploymentInputDialog.ts:118]: Amir can you share a pointer to where it says it is deprecated? From the official typescript doc, I see that both are acceptable and equivalent.
https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

The newer 'as' was created to be friendly with jsx so it works in all environments.

**Review by ranasaria — CHANGES_REQUESTED** [review]
My only concern is really about using 'type' to pick the notebook that will be used. Imagine scenario where a user wants to have two python notebooks, one is interactive version and another completely handsoff background execution version. Using a 'tag' or description field to allow user to pick the desired notebook would be more appropriate.

File: extensions/resource-deployment/src/interfaces.ts [hunk 3]
```diff
@@ -298,6 +298,14 @@ export interface NotebookPathInfo {
 	linux: string;
 }
 
+export interface NotebookInfo {
+	/**
+	 * Type of the notebook, for example: powershell, python...
+	 */
+	type: string;
```
[the remark this answers is not in the document]
Comment (alanrenmsft) [reply extensions/resource-deployment/src/interfaces.ts:305]: 1. the scenario you mentioned "Imagine scenario where a user wants to have two python notebooks, one is interactive version and another completely handsoff background execution version" is not a real world scenario, if the scenario is for background execution, there is no need to provide 2 notebooks. that should be dev-time error.

**Review by aaomidi — APPROVED** [review]

**Pull request merged**